Jcs work

gitjcs opened the pull request:
Ready for use

clee-jana commented on find_emails.py:
```diff
@@ -0,0 +1,80 @@
+from selenium.webdriver.common.by import By
+
+
+def scrape_emails(driver, domain, url, visited=[],
```
this method is about 40 lines long.  do you think there's a way to refactor it into smaller methods?

gitjcs replied:
Good idea. We could safely separate the scraping of email addresses and new urls, however let's make sure not to duplicate driver.get(url) as that is the bottleneck.

clee-jana commented on find_emails.py:
```diff
@@ -0,0 +1,80 @@
+    """Generates a list of emails found on a certain domain"""
+    if url in visited or depth > max_depth:
+        return
+    visited.append(url)
```
do you think visited should be a set?

gitjcs replied:
Absolutely.

clee-jana commented on find_emails.py:
```diff
@@ -0,0 +1,80 @@
+    visited.append(url)
+
+    driver.get(url)
+    time.sleep(1)  # Let the page load
```
huh, i thought the driver didn't return until the page was loaded?

gitjcs replied:
The driver will wait for a page to load by default, however it doesn't wait for loading of ajax requests. Given this is a generic crawler, I added the 1sec delay as precaution. It might be unnecessary on some sites and not enough on others.

When scraping specific sites I like to use selenium's WebDriverWait, which allows you to specific which element to wait for. (usually tag_name=table)

clee-jana commented on find_emails.py:
```diff
@@ -0,0 +1,80 @@
+            # Look for links on current domain
+            elif scheme.startswith("http") and netloc == domain:
+                # Clean url to account for extra / and such
+                clean_url = urllib.parse.urlunparse(url_attrs)
```
can you explain what cases you are handling here?

gitjcs replied:
selenium takes care of relative paths, so all we have to do is handle the schemes. 

scheme.startswith("mailto") will find all email addresses. could make it cleaner by replacing this with scheme == 'mailto'

scheme.startswith("http") will accept both http and https. 
netloc == domain ensures we remain on the same domain. To accept subdomains, we could easily change this to "domain in netloc"

clee-jana replied:
ah interesting, i did not know that selenium could take a relative path.

what do you think we should do about paths with a # in them?  `http://foo.com/page#comments` for example

clee-jana commented:
Thanks Julien,

I left a few comments on the PR.  This would be a similar style to how we do code reviews at Jana. We like to discuss the code, but try not to attack the individual. Please accept my comments as construction feedback and questions that we can discuss your line of thinking on.  We can go back and forth on this a little bit and you can revise your code if necessary.

Please let me via email so I know when you have finished replying/commenting, so I know to come back to this.  Just in case I don't get a github notification.

gitjcs commented:
Created new methods for scraping emails and sites and added handling for fragment identifiers so we don't unnecessarily check the same site twice. Also converted visited and emails lists to sets.

clee-jana commented on find_emails.py:
```diff
@@ -0,0 +1,104 @@
+from selenium.webdriver.common.by import By
+
+
+def scrape_emails(driver):
```
thanks! this looks much cleaner now.

clee-jana commented on find_emails.py:
```diff
@@ -0,0 +1,104 @@
+    # Yield new emails that we find
+    for email in scrape_emails(driver):
+        if email not in emails:
+            yield email
```
i like the use of the generator here.
